internal/graph/check.go

```diff
@@ -945,6 +953,499 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
+type recursiveMatchUserUsersetFunc func(ctx context.Context,
+	req *ResolveCheckRequest,
+	commonParameters *recursiveMatchUserUsersetInfo,
+	level int) (*ResolveCheckResponse, error)
+
+func parallelizeRecursiveMatchUserUserset(ctx context.Context,
+	usersetItems []string,
+	req *ResolveCheckRequest,
+	commonParameters *recursiveMatchUserUsersetInfo,
+	recursiveFunc recursiveMatchUserUsersetFunc,
+	level int,
+) (*ResolveCheckResponse, error) {
+	checkOutcomeChan := make(chan checkOutcome, commonParameters.maxConcurrentReads)
+
+	pool := concurrency.NewPool(ctx, commonParameters.maxConcurrentReads)
+	cancellableCtx, cancelFunc := context.WithCancel(ctx)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	defer func() {
+		cancelFunc()
+		wg.Wait()
+	}()
+
+	go func() {
+		defer func() {
+			_ = pool.Wait()
+			close(checkOutcomeChan)
+			wg.Done()
+		}()
+		for _, usersetItem := range usersetItems {
+			_, visited := commonParameters.visitedUserset.LoadOrStore(usersetItem, struct{}{})
+			if !visited {
+				newReq := req.clone()
+				newTupleKey := tuple.NewTupleKey(usersetItem, req.GetTupleKey().GetRelation(), req.GetTupleKey().GetUser())
+				newReq.TupleKey = newTupleKey
+				pool.Go(func(ctx context.Context) error {
+					result, err := recursiveFunc(cancellableCtx, newReq, commonParameters, level)
+					concurrency.TrySendThroughChannel(cancellableCtx, checkOutcome{
+						resp: result,
+						err:  err,
+					}, checkOutcomeChan)
+
+					return nil
+				})
+			}
+			// Note that visited does not necessary means that there are cycles.  For the following model,
+			// type user
+			// type group
+			//   relations
+			//     define member: [user, group#member]
+			// We have something like
+			// group:1#member@group:2#member
+			// group:1#member@group:3#member
+			// group:2#member@group:a#member
+			// group:3#member@group:a#member
+			// Note that both group:2#member and group:3#member has group:a#member. However, they are not cycles.
+		}
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, nil
+		case msg, ok := <-checkOutcomeChan:
+			if !ok {
+				return &ResolveCheckResponse{
+					Allowed: false,
+					ResolutionMetadata: &ResolveCheckResponseMetadata{
+						DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + commonParameters.dsCount.Load(),
+					},
+				}, nil
+			}
+			if msg.err != nil && !errors.Is(msg.err, context.Canceled) && !errors.Is(msg.err, context.DeadlineExceeded) {
+				return nil, msg.err
+			}
+			if msg.resp != nil && msg.resp.GetAllowed() {
+				return msg.resp, nil
+			}
+		}
+	}
+}
+
+func recursiveMatchUserUserset(ctx context.Context,
+	req *ResolveCheckRequest,
+	commonParameters *recursiveMatchUserUsersetInfo,
+	level int) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "recursiveMatchUserUserset")
+	defer span.End()
+
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+	if level > int(req.GetRequestMetadata().Depth) {
```

Comment:
FYI This would be simpler if you didn't need a `level` variable and `depth` started at 0 and we exited when `req.GetRequestMetadata().Depth == maxAllowedDepth` 😢  https://github.com/openfga/openfga/issues/994